# panicker3000/run_scanner.py
"""
–ì–õ–ê–í–ù–´–ô –°–ö–†–ò–ü–¢ –ó–ê–ü–£–°–ö–ê –°–ò–°–¢–ï–ú–´ ¬´–ü–ê–ù–ò–ö–Å–† 3000¬ª

–û—Å–Ω–æ–≤–Ω—ã–µ —Ñ—É–Ω–∫—Ü–∏–∏:
1. –ü—Ä–æ–≤–µ—Ä–∫–∞ –≤—Ä–µ–º–µ–Ω–∏ —Ä–∞–±–æ—Ç—ã –±–∏—Ä–∂–∏ –ú–ú–í–ë —Å —É—á—ë—Ç–æ–º –ø—Ä–∞–∑–¥–Ω–∏–∫–æ–≤
2. –ó–∞–ø—É—Å–∫ gRPC —Å–µ—Ä–≤–µ—Ä–∞ –≤ –æ—Ç–¥–µ–ª—å–Ω–æ–º –ø—Ä–æ—Ü–µ—Å—Å–µ
3. –ó–∞–ø—É—Å–∫ Telegram –±–æ—Ç–∞
4. –û–±—Ä–∞–±–æ—Ç–∫–∞ –æ—à–∏–±–æ–∫ –∏ graceful shutdown

–ò—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω–∏–µ:
    python run_scanner.py
"""

# ============================================================================
# –ù–ê–°–¢–†–û–ô–ö–ê –õ–û–ì–ì–ò–†–û–í–ê–ù–ò–Ø
# ============================================================================
import logging
import sys

logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
    handlers=[
        logging.StreamHandler(sys.stdout)
    ],
    force=True
)

# –°–Ω–∏–∂–∞–µ–º —É—Ä–æ–≤–µ–Ω—å –¥–ª—è —à—É–º–Ω—ã—Ö –±–∏–±–ª–∏–æ—Ç–µ–∫
logging.getLogger('grpc').setLevel(logging.WARNING)
logging.getLogger('urllib3').setLevel(logging.WARNING)
logging.getLogger('tinkoff').setLevel(logging.WARNING)

# ============================================================================
# –ò–ú–ü–û–†–¢–´
# ============================================================================
import os
import sys
import subprocess
import time
from datetime import datetime, time as dt_time
from typing import Optional
import signal
import atexit
import yaml
import grpc
from typing import List, Dict, Any

# –î–æ–±–∞–≤–ª—è–µ–º —Ç–µ–∫—É—â—É—é –¥–∏—Ä–µ–∫—Ç–æ—Ä–∏—é –≤ –ø—É—Ç—å
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

# –ò–º–ø–æ—Ä—Ç–∏—Ä—É–µ–º MarketCalendar –¥–ª—è –ø—Ä–æ–≤–µ—Ä–∫–∏ –≤—Ä–µ–º–µ–Ω–∏ –±–∏—Ä–∂–∏
from data.market_calendar import get_market_calendar

# ============================================================================
# –ì–õ–û–ë–ê–õ–¨–ù–´–ô –õ–û–ì–ì–ï–†
# ============================================================================
logger = logging.getLogger(__name__)

# ============================================================================
# –ö–û–ù–°–¢–ê–ù–¢–´
# ============================================================================
GRPC_SERVER_PORT = 50051
CHECK_INTERVAL = 60  # –ü—Ä–æ–≤–µ—Ä–∫–∞ –∫–∞–∂–¥—ã–µ 60 —Å–µ–∫—É–Ω–¥
SCAN_INTERVAL = 60  # –ò–Ω—Ç–µ—Ä–≤–∞–ª —Å–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏—è (—Å–µ–∫—É–Ω–¥—ã)
SCAN_COOLDOWN = 300  # –û–∂–∏–¥–∞–Ω–∏–µ –ø—Ä–∏ –∑–∞–∫—Ä—ã—Ç–æ–π –±–∏—Ä–∂–µ (—Å–µ–∫—É–Ω–¥—ã)
REQUEST_DELAY = 0.2  # –ó–∞–¥–µ—Ä–∂–∫–∞ –º–µ–∂–¥—É –∑–∞–ø—Ä–æ—Å–∞–º–∏ –∫ API


# ============================================================================
# –£–î–ê–õ–ï–ù –ö–õ–ê–°–° MarketTimeChecker - –∑–∞–º–µ–Ω—ë–Ω –Ω–∞ MarketCalendar
# ============================================================================

# ============================================================================
# –ö–õ–ê–°–° ConfigLoader
# ============================================================================
class ConfigLoader:
    """–ó–∞–≥—Ä—É–∑—á–∏–∫ –∫–æ–Ω—Ñ–∏–≥—É—Ä–∞—Ü–∏–∏"""

    @staticmethod
    def load_tickers() -> List[str]:
        """–ó–∞–≥—Ä—É–∑–∫–∞ —Å–ø–∏—Å–∫–∞ —Ç–∏–∫–µ—Ä–æ–≤ –∏–∑ –∫–æ–Ω—Ñ–∏–≥–∞"""
        try:
            config_path = os.path.join("config", "tickers.yaml")
            if not os.path.exists(config_path):
                logger.warning(f"‚ö†Ô∏è  –§–∞–π–ª –∫–æ–Ω—Ñ–∏–≥–∞ –Ω–µ –Ω–∞–π–¥–µ–Ω: {config_path}")
                return ["SBER", "GAZP", "LKOH", "GMKN", "YNDX"]

            with open(config_path, 'r', encoding='utf-8') as f:
                config = yaml.safe_load(f)

            # –ü–æ–ª—É—á–∞–µ–º —Å–ø–∏—Å–æ–∫ —Ç–∏–∫–µ—Ä–æ–≤ –∏–∑ —Å—Ç—Ä—É–∫—Ç—É—Ä—ã config
            tickers = []
            for item in config.get('tickers', []):
                if isinstance(item, dict) and 'ticker' in item:
                    tickers.append(item['ticker'])
                elif isinstance(item, str):
                    tickers.append(item)

            if not tickers:
                return ["SBER", "GAZP", "LKOH", "GMKN", "YNDX"]

            logger.info(f"üìã –ó–∞–≥—Ä—É–∂–µ–Ω–æ {len(tickers)} —Ç–∏–∫–µ—Ä–æ–≤ –∏–∑ –∫–æ–Ω—Ñ–∏–≥–∞")
            return tickers

        except Exception as e:
            logger.error(f"‚ùå –û—à–∏–±–∫–∞ –∑–∞–≥—Ä—É–∑–∫–∏ –∫–æ–Ω—Ñ–∏–≥–∞: {e}")
            return ["SBER", "GAZP", "LKOH", "GMKN", "YNDX"]

    @staticmethod
    def load_scan_settings() -> Dict[str, Any]:
        """–ó–∞–≥—Ä—É–∑–∫–∞ –Ω–∞—Å—Ç—Ä–æ–µ–∫ —Å–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏—è"""
        return {
            'scan_interval': SCAN_INTERVAL,
            'cooldown_closed': SCAN_COOLDOWN,
            'request_delay': REQUEST_DELAY,
            'max_retries': 3,
        }


# ============================================================================
# –ö–õ–ê–°–° PanickerScanner
# ============================================================================
class PanickerScanner:
    """–û—Å–Ω–æ–≤–Ω–æ–π –∫–ª–∞—Å—Å –¥–ª—è –∑–∞–ø—É—Å–∫–∞ –∏ —É–ø—Ä–∞–≤–ª–µ–Ω–∏—è —Å–∏—Å—Ç–µ–º–æ–π"""

    def __init__(self):
        self.grpc_process: Optional[subprocess.Popen] = None
        self.bot_process: Optional[subprocess.Popen] = None
        self.is_running = False
        self.scanning = False  # –§–ª–∞–≥ —Å–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏—è
        self.tickers = []  # –°–ø–∏—Å–æ–∫ —Ç–∏–∫–µ—Ä–æ–≤
        self.scan_settings = {}  # –ù–∞—Å—Ç—Ä–æ–π–∫–∏ —Å–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏—è

        # –ò–Ω–∏—Ü–∏–∞–ª–∏–∑–∏—Ä—É–µ–º MarketCalendar
        self.market_calendar = get_market_calendar()

        # –ò–ù–ò–¶–ò–ê–õ–ò–ó–ê–¶–ò–Ø –ö–û–ú–ü–û–ù–ï–ù–¢–û–í:
        self._init_components()

        # –†–µ–≥–∏—Å—Ç—Ä–∏—Ä—É–µ–º –æ–±—Ä–∞–±–æ—Ç—á–∏–∫–∏ –∑–∞–≤–µ—Ä—à–µ–Ω–∏—è
        atexit.register(self.cleanup)
        signal.signal(signal.SIGINT, self._signal_handler)
        signal.signal(signal.SIGTERM, self._signal_handler)

        logger.info("ü§ñ PanickerScanner –∏–Ω–∏—Ü–∏–∞–ª–∏–∑–∏—Ä–æ–≤–∞–Ω —Å MarketCalendar")

    def start_auto_scanning(self):
        """–ó–∞–ø—É—Å–∫ –∞–≤—Ç–æ—Å–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏—è –≤ –æ—Ç–¥–µ–ª—å–Ω–æ–º –ø–æ—Ç–æ–∫–µ"""
        if self.scanning:
            logger.warning("‚ö†Ô∏è  –ê–≤—Ç–æ—Å–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏–µ —É–∂–µ –∑–∞–ø—É—â–µ–Ω–æ")
            return

        import threading
        self.scanning = True
        scan_thread = threading.Thread(target=self._scan_loop, daemon=True)
        scan_thread.start()
        logger.info("üîÑ –ê–≤—Ç–æ—Å–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏–µ –∑–∞–ø—É—â–µ–Ω–æ –≤ –æ—Ç–¥–µ–ª—å–Ω–æ–º –ø–æ—Ç–æ–∫–µ")

    def stop_auto_scanning(self):
        """–û—Å—Ç–∞–Ω–æ–≤–∫–∞ –∞–≤—Ç–æ—Å–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏—è"""
        self.scanning = False
        logger.info("‚èπÔ∏è  –ê–≤—Ç–æ—Å–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏–µ –æ—Å—Ç–∞–Ω–æ–≤–ª–µ–Ω–æ")

    def _scan_loop(self):
        """–û—Å–Ω–æ–≤–Ω–æ–π —Ü–∏–∫–ª —Å–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏—è"""
        logger.info("üîç –ù–∞—á–∏–Ω–∞–µ–º —Ü–∏–∫–ª–∏—á–µ—Å–∫–æ–µ —Å–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏–µ —Ä—ã–Ω–∫–∞...")

        scan_count = 0
        last_scan_time = datetime.now()

        while self.scanning and self.is_running:
            try:
                current_time = datetime.now()
                time_since_last_scan = (current_time - last_scan_time).total_seconds()

                # –ü—Ä–æ–≤–µ—Ä—è–µ–º, –ø–æ—Ä–∞ –ª–∏ —Å–∫–∞–Ω–∏—Ä–æ–≤–∞—Ç—å
                if time_since_last_scan < self.scan_settings['scan_interval']:
                    time.sleep(1)
                    continue

                # –û–±–Ω–æ–≤–ª—è–µ–º –≤—Ä–µ–º—è –ø–æ—Å–ª–µ–¥–Ω–µ–≥–æ —Å–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏—è
                last_scan_time = current_time
                scan_count += 1

                logger.info(f"üîç –°–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏–µ #{scan_count} –Ω–∞—á–∞—Ç–æ –≤ {current_time.strftime('%H:%M:%S')}")

                # –í—ã–ø–æ–ª–Ω—è–µ–º —Å–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏–µ
                signals_found = self._scan_market()

                if signals_found:
                    logger.info(f"‚úÖ –ù–∞–π–¥–µ–Ω–æ {signals_found} —Å–∏–≥–Ω–∞–ª–æ–≤")
                else:
                    logger.info("‚úÖ –°–∏–≥–Ω–∞–ª–æ–≤ –Ω–µ –æ–±–Ω–∞—Ä—É–∂–µ–Ω–æ")

                # –ñ–¥—ë–º –ø–µ—Ä–µ–¥ —Å–ª–µ–¥—É—é—â–∏–º —Å–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏–µ–º
                time.sleep(1)

            except KeyboardInterrupt:
                break
            except Exception as e:
                logger.error(f"‚ùå –û—à–∏–±–∫–∞ –≤ —Ü–∏–∫–ª–µ —Å–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏—è: {e}")
                time.sleep(5)  # –ü–∞—É–∑–∞ –ø—Ä–∏ –æ—à–∏–±–∫–µ

    def _scan_market(self) -> int:
        """–°–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏–µ —Ä—ã–Ω–∫–∞ –∏ –ø–æ–∏—Å–∫ —Å–∏–≥–Ω–∞–ª–æ–≤"""
        # –ò—Å–ø–æ–ª—å–∑—É–µ–º MarketCalendar –¥–ª—è –ø—Ä–æ–≤–µ—Ä–∫–∏ –≤—Ä–µ–º–µ–Ω–∏ —Ä–∞–±–æ—Ç—ã –±–∏—Ä–∂–∏
        is_open, reason = self.market_calendar.is_market_open_now()
        if not is_open:
            logger.info(f"‚è∞ –ë–∏—Ä–∂–∞ –∑–∞–∫—Ä—ã—Ç–∞ ({reason}), –ø—Ä–æ–ø—É—Å–∫–∞–µ–º —Å–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏–µ")
            return 0

        signals_found = 0

        for ticker in self.tickers:
            if not self.scanning or not self.is_running:
                break

            try:
                logger.debug(f"üìä –°–∫–∞–Ω–∏—Ä—É–µ–º {ticker}...")

                # –í—ã–∑—ã–≤–∞–µ–º gRPC —Å–µ—Ä–≤–∏—Å –¥–ª—è —Å–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏—è —Ç–∏–∫–µ—Ä–∞
                try:
                    import grpc
                    from grpc_service.proto.generated import panicker_pb2, panicker_pb2_grpc

                    # –°–æ–∑–¥–∞—ë–º gRPC –∫–∞–Ω–∞–ª
                    channel = grpc.insecure_channel(f'localhost:{GRPC_SERVER_PORT}')
                    stub = panicker_pb2_grpc.PanickerServiceStub(channel)

                    # –í—ã–∑—ã–≤–∞–µ–º –º–µ—Ç–æ–¥ ScanTicker
                    request = panicker_pb2.ScanTickerRequest(ticker=ticker)
                    response = stub.ScanTicker(request)

                    if response.signal_found:
                        signal_data = {
                            'ticker': ticker,
                            'level': response.level,
                            'signal_type': response.signal_type,
                            'rsi': response.rsi,
                            'volume_ratio': response.volume_ratio,
                            'timestamp': datetime.now().isoformat(),
                            'message': f"{response.level} {response.signal_type} –≤ {ticker}! RSI={response.rsi:.1f}"
                        }
                    else:
                        signal_data = None

                except Exception as e:
                    logger.error(f"‚ùå –û—à–∏–±–∫–∞ gRPC —Å–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏—è {ticker}: {e}")
                    signal_data = None

                if signal_data:
                    signals_found += 1
                    self._send_alert(signal_data)

                # –ü–∞—É–∑–∞ –º–µ–∂–¥—É –∑–∞–ø—Ä–æ—Å–∞–º–∏
                time.sleep(self.scan_settings['request_delay'])

            except Exception as e:
                logger.error(f"‚ùå –û—à–∏–±–∫–∞ —Å–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏—è {ticker}: {e}")

        return signals_found

    def _send_alert(self, signal_data):
        """–û—Ç–ø—Ä–∞–≤–∫–∞ –æ–ø–æ–≤–µ—â–µ–Ω–∏—è —á–µ—Ä–µ–∑ gRPC"""
        try:
            logger.info(f"üö® –û—Ç–ø—Ä–∞–≤–∫–∞ –æ–ø–æ–≤–µ—â–µ–Ω–∏—è: {signal_data['message']}")

            # –û—Ç–ø—Ä–∞–≤–ª—è–µ–º —á–µ—Ä–µ–∑ gRPC
            try:
                import grpc
                from grpc_service.proto.generated import panicker_pb2, panicker_pb2_grpc

                channel = grpc.insecure_channel(f'localhost:{GRPC_SERVER_PORT}')
                stub = panicker_pb2_grpc.SignalsServiceStub(channel)

                # –°–æ–∑–¥–∞—ë–º –∑–∞–ø—Ä–æ—Å
                request = panicker_pb2.SendAlertRequest(
                    ticker=signal_data['ticker'],
                    level=signal_data['level'],
                    signal_type=signal_data['signal_type'],
                    message=signal_data['message'],
                    rsi=signal_data['rsi'],
                    volume_ratio=signal_data['volume_ratio']
                )

                response = stub.SendAlert(request)

                if response.success:
                    logger.info(f"üì§ –û–ø–æ–≤–µ—â–µ–Ω–∏–µ –æ—Ç–ø—Ä–∞–≤–ª–µ–Ω–æ —á–µ—Ä–µ–∑ gRPC: {signal_data['ticker']}")
                else:
                    logger.warning(f"‚ö†Ô∏è  –û–ø–æ–≤–µ—â–µ–Ω–∏–µ –Ω–µ –æ—Ç–ø—Ä–∞–≤–ª–µ–Ω–æ: {response.error_message}")

            except Exception as grpc_error:
                logger.error(f"‚ùå –û—à–∏–±–∫–∞ gRPC –æ—Ç–ø—Ä–∞–≤–∫–∏ –æ–ø–æ–≤–µ—â–µ–Ω–∏—è: {grpc_error}")
                # –õ–æ–≥–∏—Ä—É–µ–º –∫–∞–∫ –∑–∞–ø–∞—Å–Ω–æ–π –≤–∞—Ä–∏–∞–Ω—Ç
                logger.info(f"üì§ –û–ø–æ–≤–µ—â–µ–Ω–∏–µ –∑–∞–ª–æ–≥–∏—Ä–æ–≤–∞–Ω–æ: {signal_data['ticker']} - {signal_data['level']}")

        except Exception as e:
            logger.error(f"‚ùå –û—à–∏–±–∫–∞ –æ—Ç–ø—Ä–∞–≤–∫–∏ –æ–ø–æ–≤–µ—â–µ–Ω–∏—è: {e}")

    # ------------------------------------------------------------------------
    # –ò–ù–ò–¶–ò–ê–õ–ò–ó–ê–¶–ò–Ø –ö–û–ú–ü–û–ù–ï–ù–¢–û–í
    # ------------------------------------------------------------------------
    def _init_components(self):
        """–ò–Ω–∏—Ü–∏–∞–ª–∏–∑–∞—Ü–∏—è –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–æ–≤ –¥–ª—è –∞–≤—Ç–æ—Å–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏—è"""
        try:
            # –ó–∞–≥—Ä—É–∂–∞–µ–º –∫–æ–Ω—Ñ–∏–≥—É—Ä–∞—Ü–∏—é
            self.tickers = ConfigLoader.load_tickers()
            self.scan_settings = ConfigLoader.load_scan_settings()

            logger.info(f"üìã –ó–∞–≥—Ä—É–∂–µ–Ω–æ {len(self.tickers)} —Ç–∏–∫–µ—Ä–æ–≤ –¥–ª—è —Å–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏—è")
            logger.info(f"‚öôÔ∏è  –ù–∞—Å—Ç—Ä–æ–π–∫–∏ —Å–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏—è: –∏–Ω—Ç–µ—Ä–≤–∞–ª {self.scan_settings['scan_interval']}—Å")

        except Exception as e:
            logger.error(f"‚ùå –û—à–∏–±–∫–∞ –∏–Ω–∏—Ü–∏–∞–ª–∏–∑–∞—Ü–∏–∏ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–æ–≤: {e}")
            # –†–∞–±–æ—Ç–∞–µ–º —Å —Ç–µ—Å—Ç–æ–≤—ã–º–∏ –¥–∞–Ω–Ω—ã–º–∏
            self.tickers = ["SBER", "GAZP", "LKOH", "GMKN", "YNDX"]
            self.scan_settings = {
                'scan_interval': SCAN_INTERVAL,
                'cooldown_closed': SCAN_COOLDOWN,
                'request_delay': REQUEST_DELAY,
                'max_retries': 3,
            }

    # ------------------------------------------------------------------------
    # –ó–ê–ü–£–°–ö –ö–û–ú–ü–û–ù–ï–ù–¢–û–í
    # ------------------------------------------------------------------------
    def start_grpc_server(self) -> bool:
        """–ó–∞–ø—É—Å–∫ gRPC —Å–µ—Ä–≤–µ—Ä–∞ –≤ –æ—Ç–¥–µ–ª—å–Ω–æ–º –ø—Ä–æ—Ü–µ—Å—Å–µ"""
        try:
            logger.info("üöÄ –ó–∞–ø—É—Å–∫ gRPC —Å–µ—Ä–≤–µ—Ä–∞...")

            # –ö–æ–º–∞–Ω–¥–∞ –¥–ª—è –∑–∞–ø—É—Å–∫–∞ —Å–µ—Ä–≤–µ—Ä–∞
            cmd = [sys.executable, "grpc_service/grpc_server.py"]

            # –ó–∞–ø—É—Å–∫–∞–µ–º –ø—Ä–æ—Ü–µ—Å—Å
            self.grpc_process = subprocess.Popen(
                cmd,
                stdout=subprocess.PIPE,
                stderr=subprocess.STDOUT,
                bufsize=1,
                universal_newlines=False
            )

            # –î–∞—ë–º –≤—Ä–µ–º—è –Ω–∞ –∑–∞–ø—É—Å–∫
            time.sleep(3)

            # –ü—Ä–æ–≤–µ—Ä—è–µ–º, —á—Ç–æ –ø—Ä–æ—Ü–µ—Å—Å –∑–∞–ø—É—â–µ–Ω
            if self.grpc_process.poll() is not None:
                # –ü—Ä–æ—Ü–µ—Å—Å –∑–∞–≤–µ—Ä—à–∏–ª—Å—è
                stdout, _ = self.grpc_process.communicate()
                logger.error(f"‚ùå gRPC —Å–µ—Ä–≤–µ—Ä –∑–∞–≤–µ—Ä—à–∏–ª—Å—è —Å –æ—à–∏–±–∫–æ–π:\n{stdout}")
                return False

            logger.info(f"‚úÖ gRPC —Å–µ—Ä–≤–µ—Ä –∑–∞–ø—É—â–µ–Ω (PID: {self.grpc_process.pid})")
            logger.info(f"   üì° –ü–æ—Ä—Ç: {GRPC_SERVER_PORT}")

            # –ß–∏—Ç–∞–µ–º –≤—ã–≤–æ–¥ –≤ –æ—Ç–¥–µ–ª—å–Ω–æ–º –ø–æ—Ç–æ–∫–µ (—É–ø—Ä–æ—â—ë–Ω–Ω–æ)
            self._log_process_output(self.grpc_process, "gRPC Server")

            return True

        except Exception as e:
            logger.error(f"‚ùå –û—à–∏–±–∫–∞ –∑–∞–ø—É—Å–∫–∞ gRPC —Å–µ—Ä–≤–µ—Ä–∞: {e}")
            return False

    def start_telegram_bot(self) -> bool:
        """–ó–∞–ø—É—Å–∫ Telegram –±–æ—Ç–∞ –≤ –æ—Ç–¥–µ–ª—å–Ω–æ–º –ø—Ä–æ—Ü–µ—Å—Å–µ"""
        try:
            logger.info("ü§ñ –ó–∞–ø—É—Å–∫ Telegram –±–æ—Ç–∞...")

            # –ö–æ–º–∞–Ω–¥–∞ –¥–ª—è –∑–∞–ø—É—Å–∫–∞ –±–æ—Ç–∞
            cmd = [sys.executable, "bot/telegram_panicker.py"]

            # –ó–∞–ø—É—Å–∫–∞–µ–º –ø—Ä–æ—Ü–µ—Å—Å
            self.bot_process = subprocess.Popen(
                cmd,
                stdout=subprocess.PIPE,
                stderr=subprocess.STDOUT,
                bufsize=1,
                universal_newlines=False
            )

            # –î–∞—ë–º –≤—Ä–µ–º—è –Ω–∞ –∑–∞–ø—É—Å–∫
            time.sleep(5)

            # –ü—Ä–æ–≤–µ—Ä—è–µ–º, —á—Ç–æ –ø—Ä–æ—Ü–µ—Å—Å –∑–∞–ø—É—â–µ–Ω
            if self.bot_process.poll() is not None:
                # –ü—Ä–æ—Ü–µ—Å—Å –∑–∞–≤–µ—Ä—à–∏–ª—Å—è
                try:
                    stdout, _ = self.bot_process.communicate(timeout=1)
                    if stdout:
                        try:
                            decoded_stdout = stdout.decode('utf-8')
                        except UnicodeDecodeError:
                            decoded_stdout = stdout.decode('cp1251', errors='ignore')
                        logger.error(f"‚ùå Telegram –±–æ—Ç –∑–∞–≤–µ—Ä—à–∏–ª—Å—è —Å –æ—à–∏–±–∫–æ–π:\n{decoded_stdout}")
                    else:
                        logger.error("‚ùå Telegram –±–æ—Ç –∑–∞–≤–µ—Ä—à–∏–ª—Å—è –±–µ–∑ –≤—ã–≤–æ–¥–∞")
                except Exception as e:
                    logger.error(f"‚ùå –û—à–∏–±–∫–∞ –ø—Ä–∏ —á—Ç–µ–Ω–∏–∏ –≤—ã–≤–æ–¥–∞ –±–æ—Ç–∞: {e}")
                return False

            logger.info(f"‚úÖ Telegram –±–æ—Ç –∑–∞–ø—É—â–µ–Ω (PID: {self.bot_process.pid})")

            # –ß–∏—Ç–∞–µ–º –≤—ã–≤–æ–¥ –≤ –æ—Ç–¥–µ–ª—å–Ω–æ–º –ø–æ—Ç–æ–∫–µ
            self._log_process_output(self.bot_process, "Telegram Bot")

            return True

        except Exception as e:
            logger.error(f"‚ùå –û—à–∏–±–∫–∞ –∑–∞–ø—É—Å–∫–∞ Telegram –±–æ—Ç–∞: {e}")
            self.bot_process = None  # –ì–∞—Ä–∞–Ω—Ç–∏—Ä—É–µ–º, —á—Ç–æ –Ω–µ –æ—Å—Ç–∞–Ω–µ—Ç—Å—è None
            return False

    def _log_process_output(self, process: subprocess.Popen, process_name: str):
        """–ß—Ç–µ–Ω–∏–µ –≤—ã–≤–æ–¥–∞ –ø—Ä–æ—Ü–µ—Å—Å–∞ –≤ –æ—Ç–¥–µ–ª—å–Ω–æ–º –ø–æ—Ç–æ–∫–µ (—É–ø—Ä–æ—â—ë–Ω–Ω–æ)"""
        import threading

        def read_output():
            try:
                while True:
                    line = process.stdout.readline()
                    if not line:
                        break
                    try:
                        # –ü—Ä–æ–±—É–µ–º –¥–µ–∫–æ–¥–∏—Ä–æ–≤–∞—Ç—å –∫–∞–∫ UTF-8, –µ—Å–ª–∏ –Ω–µ –ø–æ–ª—É—á–∞–µ—Ç—Å—è - –∏—Å–ø–æ–ª—å–∑—É–µ–º cp1251 –¥–ª—è Windows
                        decoded_line = line.decode('utf-8').strip()
                    except UnicodeDecodeError:
                        try:
                            decoded_line = line.decode('cp1251').strip()
                        except:
                            decoded_line = line.decode('utf-8', errors='ignore').strip()

                    if decoded_line:
                        logger.info(f"[{process_name}] {decoded_line}")
            except Exception as e:
                logger.error(f"‚ùå –û—à–∏–±–∫–∞ —á—Ç–µ–Ω–∏—è –≤—ã–≤–æ–¥–∞ {process_name}: {e}")

        thread = threading.Thread(target=read_output, daemon=True)
        thread.start()

    # ------------------------------------------------------------------------
    # –û–°–ù–û–í–ù–û–ô –¶–ò–ö–õ
    # ------------------------------------------------------------------------
    def run(self):
        """–û—Å–Ω–æ–≤–Ω–æ–π —Ü–∏–∫–ª —Ä–∞–±–æ—Ç—ã —Å–∏—Å—Ç–µ–º—ã"""
        logger.info("=" * 60)
        logger.info("üöÄ –ó–ê–ü–£–°–ö –°–ò–°–¢–ï–ú–´ ¬´–ü–ê–ù–ò–ö–Å–† 3000¬ª")
        logger.info("=" * 60)

        self.is_running = True

        try:
            # 1. –ü—Ä–æ–≤–µ—Ä—è–µ–º –≤—Ä–µ–º—è –±–∏—Ä–∂–∏ —Å –ø–æ–º–æ—â—å—é MarketCalendar
            is_open, reason = self.market_calendar.is_market_open_now()
            if not is_open:
                logger.warning(f"‚è∞ –ë–∏—Ä–∂–∞ –∑–∞–∫—Ä—ã—Ç–∞: {reason}. –ó–∞–ø—É—Å–∫–∞–µ–º –≤ —Ä–µ–∂–∏–º–µ —Ç–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏—è...")
                # –í —Ç–µ—Å—Ç–æ–≤–æ–º —Ä–µ–∂–∏–º–µ –≤—Å—ë —Ä–∞–≤–Ω–æ –∑–∞–ø—É—Å–∫–∞–µ–º
                # –í —Ä–µ–∞–ª—å–Ω–æ–π —Å–∏—Å—Ç–µ–º–µ –∑–¥–µ—Å—å –º–æ–∂–Ω–æ –±—ã–ª–æ –±—ã –∂–¥–∞—Ç—å –æ—Ç–∫—Ä—ã—Ç–∏—è

            # 2. –ó–∞–ø—É—Å–∫–∞–µ–º gRPC —Å–µ—Ä–≤–µ—Ä
            if not self.start_grpc_server():
                logger.error("‚ùå –ù–µ —É–¥–∞–ª–æ—Å—å –∑–∞–ø—É—Å—Ç–∏—Ç—å gRPC —Å–µ—Ä–≤–µ—Ä. –ó–∞–≤–µ—Ä—à–µ–Ω–∏–µ.")
                return

            # 3. –ó–∞–ø—É—Å–∫–∞–µ–º Telegram –±–æ—Ç–∞
            if not self.start_telegram_bot():
                logger.error("‚ùå –ù–µ —É–¥–∞–ª–æ—Å—å –∑–∞–ø—É—Å—Ç–∏—Ç—å Telegram –±–æ—Ç–∞.")
                # gRPC —Å–µ—Ä–≤–µ—Ä –ø—Ä–æ–¥–æ–ª–∂–∞–µ—Ç —Ä–∞–±–æ—Ç–∞—Ç—å

            # 4. –ó–∞–ø—É—Å–∫–∞–µ–º –∞–≤—Ç–æ—Å–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏–µ - –î–û–ë–ê–í–õ–ï–ù–û
            logger.info("üîÑ –ó–∞–ø—É—Å–∫ –∞–≤—Ç–æ—Å–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏—è...")
            self.start_auto_scanning()

            # 5. –û—Å–Ω–æ–≤–Ω–æ–π —Ü–∏–∫–ª –º–æ–Ω–∏—Ç–æ—Ä–∏–Ω–≥–∞
            logger.info("‚úÖ –°–∏—Å—Ç–µ–º–∞ –ø–æ–ª–Ω–æ—Å—Ç—å—é –∑–∞–ø—É—â–µ–Ω–∞. –ú–æ–Ω–∏—Ç–æ—Ä–∏–Ω–≥ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–æ–≤...")

            while self.is_running:
                # –ü—Ä–æ–≤–µ—Ä—è–µ–º —Å–æ—Å—Ç–æ—è–Ω–∏–µ –ø—Ä–æ—Ü–µ—Å—Å–æ–≤
                self._check_processes()

                # –ü—Ä–æ–≤–µ—Ä—è–µ–º –≤—Ä–µ–º—è –±–∏—Ä–∂–∏ (—Ç–æ–ª—å–∫–æ –¥–ª—è –ª–æ–≥–∏—Ä–æ–≤–∞–Ω–∏—è)
                is_open, reason = self.market_calendar.is_market_open_now()
                if not is_open and self.scanning:
                    logger.info(f"üî¥ –ë–∏—Ä–∂–∞ –∑–∞–∫—Ä—ã—Ç–∞: {reason}. –ê–≤—Ç–æ—Å–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏–µ –≤—Ä–µ–º–µ–Ω–Ω–æ –ø—Ä–∏–æ—Å—Ç–∞–Ω–æ–≤–ª–µ–Ω–æ.")
                    # –°–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏–µ —Å–∞–º–æ –ø—Ä–∏–æ—Å—Ç–∞–Ω–æ–≤–∏—Ç—Å—è –≤ _scan_market()

                # –ñ–¥—ë–º –ø–µ—Ä–µ–¥ —Å–ª–µ–¥—É—é—â–µ–π –ø—Ä–æ–≤–µ—Ä–∫–æ–π
                time.sleep(CHECK_INTERVAL)

        except KeyboardInterrupt:
            logger.info("\nüõë –ü–æ–ª—É—á–µ–Ω —Å–∏–≥–Ω–∞–ª KeyboardInterrupt")
        except Exception as e:
            logger.error(f"‚ùå –ö—Ä–∏—Ç–∏—á–µ—Å–∫–∞—è –æ—à–∏–±–∫–∞ –≤ –æ—Å–Ω–æ–≤–Ω–æ–º —Ü–∏–∫–ª–µ: {e}")
        finally:
            self.cleanup()

    def _check_processes(self):
        """–ü—Ä–æ–≤–µ—Ä–∫–∞ —Å–æ—Å—Ç–æ—è–Ω–∏—è –∑–∞–ø—É—â–µ–Ω–Ω—ã—Ö –ø—Ä–æ—Ü–µ—Å—Å–æ–≤"""
        try:
            # –ü—Ä–æ–≤–µ—Ä—è–µ–º gRPC —Å–µ—Ä–≤–µ—Ä
            if self.grpc_process and self.grpc_process.poll() is not None:
                logger.error("‚ùå gRPC —Å–µ—Ä–≤–µ—Ä –∑–∞–≤–µ—Ä—à–∏–ª—Å—è –Ω–µ–æ–∂–∏–¥–∞–Ω–Ω–æ")
                # –ú–æ–∂–Ω–æ –ø–æ–ø—Ä–æ–±–æ–≤–∞—Ç—å –ø–µ—Ä–µ–∑–∞–ø—É—Å—Ç–∏—Ç—å
                # self.start_grpc_server()

            # –ü—Ä–æ–≤–µ—Ä—è–µ–º Telegram –±–æ—Ç–∞
            if self.bot_process and self.bot_process.poll() is not None:
                logger.error("‚ùå Telegram –±–æ—Ç –∑–∞–≤–µ—Ä—à–∏–ª—Å—è –Ω–µ–æ–∂–∏–¥–∞–Ω–Ω–æ")
                # –ü–æ–ø—Ä–æ–±—É–µ–º –ø–µ—Ä–µ–∑–∞–ø—É—Å—Ç–∏—Ç—å
                logger.info("üîÑ –ü–µ—Ä–µ–∑–∞–ø—É—Å–∫ Telegram –±–æ—Ç–∞...")
                self.start_telegram_bot()

        except Exception as e:
            logger.error(f"‚ùå –û—à–∏–±–∫–∞ –ø—Ä–æ–≤–µ—Ä–∫–∏ –ø—Ä–æ—Ü–µ—Å—Å–æ–≤: {e}")

    # ------------------------------------------------------------------------
    # –ó–ê–í–ï–†–®–ï–ù–ò–ï –†–ê–ë–û–¢–´
    # ------------------------------------------------------------------------
    def cleanup(self):
        """–ö–æ—Ä—Ä–µ–∫—Ç–Ω–æ–µ –∑–∞–≤–µ—Ä—à–µ–Ω–∏–µ –≤—Å–µ—Ö –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–æ–≤"""
        if not self.is_running:
            return

        self.is_running = False
        logger.info("üßπ –ó–∞–≤–µ—Ä—à–µ–Ω–∏–µ —Ä–∞–±–æ—Ç—ã —Å–∏—Å—Ç–µ–º—ã...")

        # –û—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ–º –∞–≤—Ç–æ—Å–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏–µ - –î–û–ë–ê–í–õ–ï–ù–û
        if self.scanning:
            logger.info("‚èπÔ∏è  –û—Å—Ç–∞–Ω–æ–≤–∫–∞ –∞–≤—Ç–æ—Å–∫–∞–Ω–∏—Ä–æ–≤–∞–Ω–∏—è...")
            self.stop_auto_scanning()
            time.sleep(1)  # –î–∞—ë–º –≤—Ä–µ–º—è –Ω–∞ –∑–∞–≤–µ—Ä—à–µ–Ω–∏–µ

        # –û—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ–º Telegram –±–æ—Ç–∞
        if self.bot_process:
            logger.info("üõë –û—Å—Ç–∞–Ω–æ–≤–∫–∞ Telegram –±–æ—Ç–∞...")
            try:
                self.bot_process.terminate()
                self.bot_process.wait(timeout=5)
                logger.info("‚úÖ Telegram –±–æ—Ç –æ—Å—Ç–∞–Ω–æ–≤–ª–µ–Ω")
            except subprocess.TimeoutExpired:
                logger.warning("‚ö†Ô∏è  Telegram –±–æ—Ç –Ω–µ –æ—Ç–≤–µ—Ç–∏–ª –Ω–∞ terminate, –ø—Ä–∏–Ω—É–¥–∏—Ç–µ–ª—å–Ω–∞—è –æ—Å—Ç–∞–Ω–æ–≤–∫–∞")
                self.bot_process.kill()
            except Exception as e:
                logger.error(f"‚ùå –û—à–∏–±–∫–∞ –æ—Å—Ç–∞–Ω–æ–≤–∫–∏ Telegram –±–æ—Ç–∞: {e}")

        # –û—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ–º gRPC —Å–µ—Ä–≤–µ—Ä
        if self.grpc_process:
            logger.info("üõë –û—Å—Ç–∞–Ω–æ–≤–∫–∞ gRPC —Å–µ—Ä–≤–µ—Ä–∞...")
            try:
                self.grpc_process.terminate()
                self.grpc_process.wait(timeout=5)
                logger.info("‚úÖ gRPC —Å–µ—Ä–≤–µ—Ä –æ—Å—Ç–∞–Ω–æ–≤–ª–µ–Ω")
            except subprocess.TimeoutExpired:
                logger.warning("‚ö†Ô∏è  gRPC —Å–µ—Ä–≤–µ—Ä –Ω–µ –æ—Ç–≤–µ—Ç–∏–ª –Ω–∞ terminate, –ø—Ä–∏–Ω—É–¥–∏—Ç–µ–ª—å–Ω–∞—è –æ—Å—Ç–∞–Ω–æ–≤–∫–∞")
                self.grpc_process.kill()
            except Exception as e:
                logger.error(f"‚ùå –û—à–∏–±–∫–∞ –æ—Å—Ç–∞–Ω–æ–≤–∫–∏ gRPC —Å–µ—Ä–≤–µ—Ä–∞: {e}")

        logger.info("üëã –°–∏—Å—Ç–µ–º–∞ ¬´–ü–∞–Ω–∏–∫—ë—Ä 3000¬ª –∑–∞–≤–µ—Ä—à–∏–ª–∞ —Ä–∞–±–æ—Ç—É")

    def _signal_handler(self, signum, frame):
        """–û–±—Ä–∞–±–æ—Ç—á–∏–∫ —Å–∏–≥–Ω–∞–ª–æ–≤ –∑–∞–≤–µ—Ä—à–µ–Ω–∏—è"""
        logger.info(f"üì° –ü–æ–ª—É—á–µ–Ω —Å–∏–≥–Ω–∞–ª {signum}")
        self.cleanup()
        sys.exit(0)


# ============================================================================
# –ö–û–ú–ê–ù–î–ù–ê–Ø –°–¢–†–û–ö–ê
# ============================================================================
def main():
    """–¢–æ—á–∫–∞ –≤—Ö–æ–¥–∞ –¥–ª—è –∑–∞–ø—É—Å–∫–∞ –∏–∑ –∫–æ–º–∞–Ω–¥–Ω–æ–π —Å—Ç—Ä–æ–∫–∏"""
    import argparse

    parser = argparse.ArgumentParser(description='–ó–∞–ø—É—Å–∫ —Å–∏—Å—Ç–µ–º—ã –ü–∞–Ω–∏–∫—ë—Ä 3000')
    parser.add_argument(
        '--test',
        action='store_true',
        help='–ó–∞–ø—É—Å–∫ –≤ —Ç–µ—Å—Ç–æ–≤–æ–º —Ä–µ–∂–∏–º–µ (–±–µ–∑ –ø—Ä–æ–≤–µ—Ä–∫–∏ –≤—Ä–µ–º–µ–Ω–∏ –±–∏—Ä–∂–∏)'
    )
    parser.add_argument(
        '--only-grpc',
        action='store_true',
        help='–ó–∞–ø—É—Å–∫ —Ç–æ–ª—å–∫–æ gRPC —Å–µ—Ä–≤–µ—Ä–∞'
    )
    parser.add_argument(
        '--only-bot',
        action='store_true',
        help='–ó–∞–ø—É—Å–∫ —Ç–æ–ª—å–∫–æ Telegram –±–æ—Ç–∞'
    )

    args = parser.parse_args()

    try:
        scanner = PanickerScanner()

        if args.only_grpc:
            logger.info("üöÄ –ó–∞–ø—É—Å–∫ —Ç–æ–ª—å–∫–æ gRPC —Å–µ—Ä–≤–µ—Ä–∞...")
            scanner.start_grpc_server()
            # –ñ–¥—ë–º –∑–∞–≤–µ—Ä—à–µ–Ω–∏—è
            while scanner.grpc_process and scanner.grpc_process.poll() is None:
                time.sleep(1)

        elif args.only_bot:
            logger.info("ü§ñ –ó–∞–ø—É—Å–∫ —Ç–æ–ª—å–∫–æ Telegram –±–æ—Ç–∞...")
            scanner.start_telegram_bot()
            # –ñ–¥—ë–º –∑–∞–≤–µ—Ä—à–µ–Ω–∏—è
            while scanner.bot_process and scanner.bot_process.poll() is None:
                time.sleep(1)

        else:
            # –ü–æ–ª–Ω—ã–π –∑–∞–ø—É—Å–∫
            scanner.run()

    except KeyboardInterrupt:
        logger.info("\nüëã –ó–∞–≤–µ—Ä—à–µ–Ω–∏–µ –ø–æ –∑–∞–ø—Ä–æ—Å—É –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—è")
    except Exception as e:
        logger.error(f"‚ùå –ö—Ä–∏—Ç–∏—á–µ—Å–∫–∞—è –æ—à–∏–±–∫–∞: {e}")
        sys.exit(1)


# ============================================================================
# –ó–ê–ü–£–°–ö
# ============================================================================
if __name__ == "__main__":
    main()